trin-core/src/utils/portal_wire.rs

```diff
@@ -0,0 +1,141 @@
+            .map_err(|err| anyhow!("unable to write to content payload buf: {err}"))?;
+    }
+    Ok(content_payload.into_inner())
+}
+
+/// Try to read up to five LEB128 bytes (The maximum content size allowed for this application is limited to `uint32`).
+pub fn read_varint(buf: &[u8]) -> anyhow::Result<(usize, u32)> {
+    for i in 1..6 {
+        match leb128::read::unsigned(&mut &buf[0..i]) {
+            Ok(varint) => {
+                let varint = u32::try_from(varint).map_err(|_| {
+                    anyhow!("Exceed maximum allowed varint value of u32 bytes size")
+                })?;
+                return Ok((i, varint));
+            }
+            Err(_) => continue,
+        }
+    }
+    Err(anyhow!("Unable to read varint index"))
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use rstest::rstest;
+
+    #[rstest]
+    #[case(u8::MIN as u32)]
+    #[case(u8::MAX as u32)]
+    #[case(u16::MAX as u32)]
+    #[case(u32::MAX)]
+    fn test_read_varint(#[case] varint: u32) {
+        let mut buf = [0; 1024];
+        let mut writable = &mut buf[..];
+
+        let bytes_written = leb128::write::unsigned(&mut writable, varint.into()).unwrap();
+
+        let (bytes_read, varint_result) = read_varint(&buf[..]).unwrap();
+
+        assert_eq!(bytes_read, bytes_written);
+        assert_eq!(varint_result, varint);
+    }
+
+    #[test]
+    #[should_panic(expected = "Unable to read varint index")]
+    fn test_read_varint_max_read_bytes() {
+        let mut buf = [0; 1024];
+        let mut writable = &mut buf[..];
+        let varint = u64::MAX;
+
+        leb128::write::unsigned(&mut writable, varint).unwrap();
+
+        read_varint(&buf[..]).unwrap();
+    }
+
+    #[test]
+    #[should_panic(expected = "Exceed maximum allowed varint value of u32 bytes size")]
+    fn test_read_varint_max_varint_size() {
+        let mut buf = [0; 1024];
+        let mut writable = &mut buf[..];
+        let varint = u32::MAX as usize + 1;
+
+        leb128::write::unsigned(&mut writable, varint as u64).unwrap();
+
+        read_varint(&buf[..]).unwrap();
+    }
+
+    #[test]
+    #[should_panic(expected = "Error reading content item: failed to fill whole buffer")]
+    fn test_decode_content_payload_corrupted() {
+        let hex_payload = "030101010201";
+        let payload = hex::decode(hex_payload).unwrap();
+        decode_content_payload(payload).unwrap();
+    }
+
+    #[test]
+    fn test_encode_decode_content_payload() {
+        let expected_content_items: Vec<Bytes> = vec![vec![1, 1].into(), vec![2, 2, 2].into()];
+
+        let content_payload = encode_content_payload(&expected_content_items)
+            .unwrap()
+            .to_vec();
+        let content_items: Vec<Bytes> = decode_content_payload(content_payload)
+            .unwrap()
+            .into_iter()
+            .map(|content| Bytes::from(content.to_vec()))
+            .collect();
+
+        assert_eq!(expected_content_items, content_items);
```

Comment:
nit: For equality assertions, I find it easier to read `assert_eq!(actual, expected)`